### PR TITLE
Forget unmanaged state in NetworkManager

### PR DESCRIPTION
**Pull request by tyll**
When removing a profile with initscripts, also notify NetworkManager so
it can forget that the specified device was to be ignored.

### CODE REVIEW

File: library/network_connections.py [hunk 1]
```diff
@@ -2340,6 +2341,36 @@ def check_name(self, idx, name=None):
...
         return f
 
+    def forget_nm_connection(self, path):
+        """
+        Forget a NetworkManager connection by loading the path of the deleted
+        profile. This inverts the effect of loading a profile with
+        `NM_CONTROLLED=no` earlier, which made NetworkManager ignore the
+        device.
+
+        This does ot use the Python libnm bindings because they might not be
```
Comment (pcahyna) [inline library/network_connections.py:2351]: *not

File: library/network_connections.py [hunk 2]
```diff
@@ -2340,6 +2341,36 @@ def check_name(self, idx, name=None):
...
+        present on the system, since the module is currently operating for the
```
Comment (pcahyna) [inline library/network_connections.py:2352]: Would it make sense to install those bindings every time that there is a NetworkManager package present on the system, even if the module is not operating for the nm provider? What package is it actually, `python-gobject-base`?

File: library/network_connections.py (hunk 1 above)
Comment (tyll) [reply library/network_connections.py]: thx, fixed

File: library/network_connections.py (hunk 2 above)
Comment (tyll) [reply library/network_connections.py:2352]: > Would it make sense to install those bindings every time that there is a NetworkManager package present on the system, even if the module is not operating for the nm provider? 

Installing it seems to be rather intrusive. This is now the same code I proposed for https://bugzilla.redhat.com/show_bug.cgi?id=1832897 therefore I do not see much harm in the current way. Another advantage of this approach is that it keeps the provider specific code separate.

> What package is it actually, `python-gobject-base`?
Yes or `python3-gobject-base`.

**Comment by pcahyna** [comment]
Is the first commit related to the second in any way?

**Comment by tyll** [comment]
> Is the first commit related to the second in any way?

The downstream bug https://bugzilla.redhat.com/show_bug.cgi?id=1800627 exposed the problem using the test from the first commit. I can also move it into its own PR If you prefer.

**Comment by pcahyna** [comment]
> > Is the first commit related to the second in any way?
> 
> The downstream bug https://bugzilla.redhat.com/show_bug.cgi?id=1800627 exposed the problem using the test from the first commit. 

I know, but the former version did expose the problem just as well as the new one, didn't it? So the change is not related to the fact that the test can expose the problem.

BTW. Why does a RHEL 7 CI test fail? The failure looks real and related to the PR.

**Comment by tyll** [comment]
> > > Is the first commit related to the second in any way?
> > 
> > 
> > The downstream bug https://bugzilla.redhat.com/show_bug.cgi?id=1800627 exposed the problem using the test from the first commit.
> 
> I know, but the former version did expose the problem just as well as the new one, didn't it? So the change is not related to the fact that the test can expose the problem.
> 
> BTW. Why does a RHEL 7 CI test fail? The failure looks real and related to the PR. EDIT: I was able to reproduce the failure locally with `rhel-guest-image-7.7-200`.

@thom311 any idea? Does NetworkManager on RHEL/CentOS 7 maybe not support forgetting individual profiles?

**Comment by pcahyna** [comment]
Now even rhel-8 is failing.

**Comment by tyll** [comment]
> Now even rhel-8 is failing.

there was a bug in the test which is why centos 7 was not failing but it should fail now similar to rhel 7.

**Comment by pcahyna** [comment]
Good to hear that CentOS 7 and RHEL 7 are at least behaving consistently.

**Comment by tyll** [comment]
> Good to hear that CentOS 7 and RHEL 7 are at least behaving consistently.

yes, the CentOS image does not contain NetworkManager but it needs to be present when using initscripts to make NM store the state. Now it should be installed if necessary before using initscripts.

**Comment by thom311** [comment]
> @thom311 any idea? Does NetworkManager on RHEL/CentOS 7 maybe not support forgetting individual profiles?

If you issue (the D-Bus equivalent of) `nmcli connection reload` or `nmcli connection load`, then files that no longer exist on disk must be unloaded/removed. The effect of that is pretty much the same as issuing `nmcli connection delete`. If that's not the case, please report a bug.

**Comment by tyll** [comment]
> > @thom311 any idea? Does NetworkManager on RHEL/CentOS 7 maybe not support forgetting individual profiles?
> 
> If you issue (the D-Bus equivalent of) `nmcli connection reload` or `nmcli connection load`, then files that no longer exist on disk must be unloaded/removed. The effect of that is pretty much the same as issuing `nmcli connection delete`. If that's not the case, please report a bug.

It works like this in RHEL/CentOS 8 but not 7.

**Comment by thom311** [comment]
> It works like this in RHEL/CentOS 8 but not 7.

Ah. You are right. That changed with the large rework in 1.20. Nowadays, "load" really means: revisit all related files on disk, and update them -- possibly remove them.

Indeed that changed since RHEL7. The rework cannot be backported, because it's too large. And there is no desire to try to come up with a "small" adjustment for this behaviour.

Sorry about the inconvenience, and that I didn't realize this to be the case right away.

**Comment by pcahyna** [comment]
Well, better to have it fixed in RHEL 8 only than to not have it fixed at all. I will adjust the affected test somehow so the problematic part is skipped on RHEL 7.

**Pull request merged**